File: app/lib/services/booking-service.ts
import {
  BookingSessionStatus,
  BookingStatus,
  PaymentStatus,
  RoomStatus,
} from "@/app/types/prisma-enums";
import type { PrismaTransactionClient } from "@/app/types/prisma";
import { ApplicationConstants, Messages } from "@/app/constants";
import { CreateBookingRequest } from "@/app/types/booking";
import {
  assertRoomAvailableForFinalBooking,
  calculateNumberOfNights,
  checkRoomAvailability,
  lockRoomForBooking,
  RoomNotAvailableError,
} from "@/app/lib/availability";
import { prisma } from "@/app/lib/prisma";
import {
  generateBookingReference,
  generatePaymentTransactionId,
} from "@/app/lib/utils/reference-number";

function validateBasicBookingRequest(request: CreateBookingRequest): string | null {
  if (!request.roomId) return Messages.Booking.RoomRequired;
  if (!request.firstName?.trim()) return Messages.Booking.FirstNameRequired;
  if (!request.lastName?.trim()) return Messages.Booking.LastNameRequired;
  if (!request.contactEmail?.trim()) return Messages.Booking.ContactEmailRequired;
  if (!request.addressLine1?.trim()) return Messages.Booking.AddressRequired;
  if (!request.city?.trim()) return Messages.Booking.CityRequired;
  if (!request.state?.trim()) return Messages.Booking.StateRequired;
  if (!request.zipCode?.trim()) return Messages.Booking.ZipCodeRequired;
  if (!request.checkInDate || !request.checkOutDate) return Messages.Booking.DatesRequired;

  return null;
}

function isValidEmail(value: string): boolean {
  return /^[^\s@]+@[^\s@]+\.[^\s@]+$/.test(value);
}

function isSafeText(value: string): boolean {
  return /^[a-zA-Z0-9\s.,'-]+$/.test(value);
}

function hasInvalidGuestName(request: CreateBookingRequest): boolean {
  return (
    !isSafeText(request.firstName) ||
    !isSafeText(request.lastName) ||
    request.guests?.some(
      (guest) => !isSafeText(guest.firstName) || !isSafeText(guest.lastName)
    ) === true
  );
}

export async function createBooking(request: CreateBookingRequest) {
  const validationMessage = validateBasicBookingRequest(request);

  if (validationMessage) {
    return { success: false, status: 400, message: validationMessage };
  }

  if (!isValidEmail(request.contactEmail.trim())) {
    return { success: false, status: 400, message: Messages.Booking.InvalidEmail };
  }

  if (hasInvalidGuestName(request)) {
    return { success: false, status: 400, message: Messages.Booking.InvalidGuestName };
  }

  const checkInDate = new Date(request.checkInDate);
  const checkOutDate = new Date(request.checkOutDate);

  if (Number.isNaN(checkInDate.getTime()) || Number.isNaN(checkOutDate.getTime())) {
    return { success: false, status: 400, message: Messages.Booking.InvalidDates };
  }

  if (checkOutDate <= checkInDate) {
    return { success: false, status: 400, message: Messages.Booking.CheckoutMustBeAfterCheckin };
  }

  const adultCount = request.adultCount;
  const childCount = request.childCount ?? 0;
  const infantCount = request.infantCount ?? 0;
  const petCount = request.petCount ?? 0;
  const totalGuestCount = adultCount + childCount + infantCount;

  if (adultCount < 1) {
    return { success: false, status: 400, message: Messages.Booking.AdultRequired };
  }

  if (childCount < 0 || infantCount < 0 || petCount < 0) {
    return { success: false, status: 400, message: Messages.Booking.InvalidGuestCount };
  }

  const room = await prisma.room.findFirst({
    where: {
      id: request.roomId,
      isActive: true,
      status: RoomStatus.AVAILABLE,
    },
  });

  if (!room) {
    return { success: false, status: 404, message: Messages.Booking.RoomNotAvailable };
  }

  if (totalGuestCount > room.maxGuests) {
    return { success: false, status: 400, message: Messages.Booking.RoomCapacityExceeded };
  }

  if (!room.petsAllowed && petCount > 0) {
    return {
      success: false,
      status: 400,
      message: Messages.Booking.PetsNotAllowed,
    };
  }

  if (room.petsAllowed && petCount > 2) {
    return {
      success: false,
      status: 400,
      message: Messages.RoomSearch.MaxPetsExceeded,
    };
  }

  let bookingSessionId: number | undefined;

  if (request.bookingSessionToken) {
    const session = await prisma.bookingSession.findUnique({
      where: { token: request.bookingSessionToken },
    });

    if (!session || session.status !== BookingSessionStatus.ACTIVE) {
      return {
        success: false,
        status: 409,
        message: Messages.Booking.InvalidBookingSession,
      };
    }

    if (session.expiresAt <= new Date()) {
      await prisma.bookingSession.update({
        where: { id: session.id },
        data: { status: BookingSessionStatus.EXPIRED },
      });

      return {
        success: false,
        status: 409,
        message: Messages.Booking.InvalidBookingSession,
      };
    }

    if (
      session.roomId !== room.id ||
      session.guestCount !== totalGuestCount ||
      session.checkInDate.getTime() !== checkInDate.getTime() ||
      session.checkOutDate.getTime() !== checkOutDate.getTime()
    ) {
      return {
        success: false,
        status: 400,
        message: Messages.Booking.InvalidBookingSession,
      };
    }

    bookingSessionId = session.id;
  }

  const availability = await checkRoomAvailability(
    room.id,
    request.checkInDate,
    request.checkOutDate
  );

  if (!availability.available) {
    return { success: false, status: 409, message: availability.message };
  }

  const numberOfNights = calculateNumberOfNights(checkInDate, checkOutDate);
  const pricePerNight = Number(room.pricePerNight);
  const discountAmount = 0;
  const taxAmount = Number(
    (pricePerNight * numberOfNights * ApplicationConstants.TaxRate).toFixed(2)
  );
  const totalPrice = Number(
    (pricePerNight * numberOfNights - discountAmount + taxAmount).toFixed(2)
  );

  try {
    const booking = await prisma.$transaction(
      async (tx: PrismaTransactionClient) => {
        await lockRoomForBooking(tx, room.id);

        await assertRoomAvailableForFinalBooking(tx, {
          roomId: room.id,
          checkInDate,
          checkOutDate,
          bookingSessionId,
        });

        const referenceNumber = generateBookingReference();
        const paymentTransactionId = generatePaymentTransactionId();

        const createdBooking = await tx.booking.create({
          data: {
            referenceNumber,
            roomId: room.id,
            bookingSessionId,
            memberId: request.memberId,
            bookingType: request.bookingType,
            firstName: request.firstName.trim(),
            lastName: request.lastName.trim(),
            gender: request.gender,
            contactEmail: request.contactEmail.trim(),
            adultCount,
            childCount,
            infantCount,
            petCount,
            checkInDate,
            checkOutDate,
            specialRequests: request.specialRequests?.trim(),
            addressLine1: request.addressLine1.trim(),
            addressLine2: request.addressLine2?.trim(),
            city: request.city.trim(),
            state: request.state.trim(),
            zipCode: request.zipCode.trim(),
            country: request.country ?? ApplicationConstants.DefaultCountry,
            pricePerNight,
            numberOfNights,
            discountAmount,
            taxAmount,
            totalPrice,
            paymentStatus: PaymentStatus.PAID,
            paymentTransactionId,
            status: BookingStatus.CONFIRMED,
            confirmationEmailSent: true,
            bookingSource: ApplicationConstants.BookingSource,
            guests: {
              create:
                request.guests?.map((guest) => ({
                  sequence: guest.sequence,
                  firstName: guest.firstName.trim(),
                  lastName: guest.lastName.trim(),
                  gender: guest.gender,
                  ageGroup: guest.ageGroup,
                })) ?? [],
            },
          },
          include: {
            room: true,
            guests: true,
          },
        });

        if (bookingSessionId) {
          await tx.bookingSession.update({
            where: { id: bookingSessionId },
            data: { status: BookingSessionStatus.CONSUMED },
          });
        }

        return createdBooking;
      },
      {
        isolationLevel: "Serializable",
        maxWait: 5000,
        timeout: 10000,
      }
    );

    console.info("Booking created successfully", {
      referenceNumber: booking.referenceNumber,
      roomId: booking.roomId,
      bookingType: booking.bookingType,
    });

    return {
      success: true,
      status: 201,
      data: {
        referenceNumber: booking.referenceNumber,
        roomName: booking.room.name,
        checkInDate: booking.checkInDate,
        checkOutDate: booking.checkOutDate,
        numberOfNights: booking.numberOfNights,
        totalPrice: Number(booking.totalPrice),
        paymentStatus: booking.paymentStatus,
        bookingStatus: booking.status,
        confirmationEmailSent: booking.confirmationEmailSent,
      },
    };
  } catch (error) {
    if (error instanceof RoomNotAvailableError) {
      return { success: false, status: error.status, message: error.message };
    }

    throw error;
  }
}

export async function getBookingByReferenceNumber(referenceNumber: string) {
  const booking = await prisma.booking.findUnique({
    where: {
      referenceNumber,
    },
    include: {
      room: true,
      member: true,
      guests: {
        orderBy: {
          sequence: "asc",
        },
      },
    },
  });

  if (!booking) {
    return {
      success: false,
      status: 404,
      message: Messages.Booking.BookingNotFound,
    };
  }

  return {
    success: true,
    status: 200,
    data: booking,
  };
}

export async function cancelBooking(referenceNumber: string) {
  const booking = await prisma.booking.findUnique({
    where: { referenceNumber },
  });

  if (!booking) {
    return {
      success: false,
      status: 404,
      message: Messages.Booking.BookingNotFound,
    };
  }

  if (booking.status === BookingStatus.CANCELLED) {
    return {
      success: false,
      status: 400,
      message: Messages.Booking.BookingAlreadyCancelled,
    };
  }

  const cancelledBooking = await prisma.booking.update({
    where: { referenceNumber },
    data: {
      status: BookingStatus.CANCELLED,
      paymentStatus: PaymentStatus.REFUNDED,
    },
  });

  console.info("Booking cancelled successfully", {
    referenceNumber: cancelledBooking.referenceNumber,
  });

  return {
    success: true,
    status: 200,
    data: {
      referenceNumber: cancelledBooking.referenceNumber,
      bookingStatus: cancelledBooking.status,
      paymentStatus: cancelledBooking.paymentStatus,
      message: Messages.Booking.BookingCancelled,
    },
  };
}

File: backend/Wynn.Booking.Application/Bookings/BookingService.cs
using Wynn.Booking.Application.Abstractions.Notifications;
using Wynn.Booking.Application.BookingSessions;
using Wynn.Booking.Application.Bookings.Dtos;
using Wynn.Booking.Application.Common;
using Wynn.Booking.Application.Rooms;
using Wynn.Booking.Domain.Entities;
using Wynn.Booking.Domain.Enums;
using Microsoft.Extensions.Logging;
using Wynn.Booking.Domain.Exceptions;
using BookingEntity = Wynn.Booking.Domain.Entities.Booking;

namespace Wynn.Booking.Application.Bookings;

public sealed class BookingService(
    IBookingRepository bookingRepository,
    IBookingSessionRepository bookingSessionRepository,
    IRoomRepository roomRepository,
    IRoomSearchService roomSearchService,
    IReservationConfirmationNotifier confirmationNotifier,
    ICurrentUserContext currentUser,
    ILogger<BookingService> logger) : IBookingService
{
    public async Task<ServiceResult<CreateBookingResponseDto>> CreateAsync(
        CreateBookingRequestDto request,
        CancellationToken cancellationToken = default)
    {
        var dateValidation = DateHelpers.ValidateStayDateRange(
            request.CheckInDate,
            request.CheckOutDate);

        if (!dateValidation.IsValid)
        {
            return ServiceResult<CreateBookingResponseDto>.Fail(
                dateValidation.ErrorMessage!,
                400);
        }

        var checkIn = DateHelpers.ParseDateOnly(request.CheckInDate)!.Value;
        var checkOut = DateHelpers.ParseDateOnly(request.CheckOutDate)!.Value;

        var childCount = request.ChildCount ?? 0;
        var infantCount = request.InfantCount ?? 0;
        var petCount = request.PetCount ?? 0;
        var totalGuestCount = request.AdultCount + childCount + infantCount;

        var room = await roomRepository.GetByIdAsync(request.RoomId, cancellationToken);

        if (room is null || !room.IsActive || room.Status != RoomStatus.Available)
        {
            return ServiceResult<CreateBookingResponseDto>.Fail(
                ApplicationMessages.Booking.SelectedRoomNotAvailable,
                404);
        }

        if (totalGuestCount > room.MaxGuests)
        {
            return ServiceResult<CreateBookingResponseDto>.Fail(
                ApplicationMessages.Booking.GuestCountExceedsCapacity,
                400);
        }

        if (!room.PetsAllowed && petCount > 0)
        {
            return ServiceResult<CreateBookingResponseDto>.Fail(
                ApplicationMessages.Booking.PetsNotAllowed,
                400);
        }

        if (room.PetsAllowed && petCount > 2)
        {
            return ServiceResult<CreateBookingResponseDto>.Fail(
                ApplicationMessages.Booking.MaxPetsExceeded,
                400);
        }

        if (string.IsNullOrWhiteSpace(request.BookingSessionToken))
        {
            return ServiceResult<CreateBookingResponseDto>.Fail(
                ApplicationMessages.Validation.BookingSessionTokenRequired,
                400);
        }

        var session = await bookingSessionRepository.GetByTokenWithRoomAsync(
            request.BookingSessionToken,
            cancellationToken);

        if (session is null || session.Status != BookingSessionStatus.Active)
        {
            return ServiceResult<CreateBookingResponseDto>.Fail(
                ApplicationMessages.Booking.SessionInvalidOrExpired,
                409);
        }

        if (session.ExpiresAt <= DateTime.UtcNow)
        {
            await bookingSessionRepository.ExpireAsync(session.Id, cancellationToken);
            return ServiceResult<CreateBookingResponseDto>.Fail(
                ApplicationMessages.Booking.SessionInvalidOrExpired,
                409);
        }

        if (session.RoomId != room.Id ||
            session.GuestCount != totalGuestCount ||
            session.CheckInDate != checkIn ||
            session.CheckOutDate != checkOut)
        {
            return ServiceResult<CreateBookingResponseDto>.Fail(
                ApplicationMessages.Booking.SessionMismatch,
                400);
        }

        var bookingSessionId = session.Id;

        var availability = await roomSearchService.CheckAvailabilityAsync(
            room.Id,
            request.CheckInDate,
            request.CheckOutDate,
            cancellationToken);

        if (availability.Data is not { Available: true })
        {
            return ServiceResult<CreateBookingResponseDto>.Fail(
                availability.Data?.Message ?? ApplicationMessages.Booking.RoomNotAvailable,
                409);
        }

        var numberOfNights = DateHelpers.CalculateNumberOfNights(checkIn, checkOut);
        var pricePerNight = room.PricePerNight;
        const decimal discountAmount = 0m;
        var taxAmount = decimal.Round(pricePerNight * numberOfNights * BookingConstants.TaxRate, 2);
        var totalPrice = decimal.Round(pricePerNight * numberOfNights - discountAmount + taxAmount, 2);
        var now = DateTime.UtcNow;

        int? memberId = null;
        var bookingType = request.BookingType;

        if (request.BookingType == BookingType.Member)
        {
            memberId = request.MemberId ?? currentUser.MemberId;

            if (!memberId.HasValue)
            {
                return ServiceResult<CreateBookingResponseDto>.Fail(
                    ApplicationMessages.Authorization.AuthenticationRequiredSignInAgain,
                    401);
            }

            var memberContactError = ValidateMemberContactMatchesAccount(request);
            if (memberContactError is not null)
            {
                return ServiceResult<CreateBookingResponseDto>.Fail(memberContactError, 400);
            }
        }

        var booking = new BookingEntity
        {
            ReferenceNumber = ReferenceGenerator.GenerateBookingReference(),
            RoomId = room.Id,
            BookingSessionId = bookingSessionId,
            MemberId = memberId,
            BookingType = bookingType,
            FirstName = request.FirstName.Trim(),
            LastName = request.LastName.Trim(),
            Gender = request.Gender,
            ContactEmail = request.ContactEmail.Trim(),
            AdultCount = request.AdultCount,
            ChildCount = childCount,
            InfantCount = infantCount,
            PetCount = petCount,
            CheckInDate = checkIn,
            CheckOutDate = checkOut,
            SpecialRequests = request.SpecialRequests?.Trim(),
            AddressLine1 = request.AddressLine1.Trim(),
            AddressLine2 = request.AddressLine2?.Trim(),
            City = request.City.Trim(),
            State = request.State.Trim(),
            ZipCode = request.ZipCode.Trim(),
            Country = request.Country ?? BookingConstants.DefaultCountry,
            PricePerNight = pricePerNight,
            NumberOfNights = numberOfNights,
            DiscountAmount = discountAmount,
            TaxAmount = taxAmount,
            TotalPrice = totalPrice,
            PaymentStatus = PaymentStatus.Paid,
            PaymentTransactionId = ReferenceGenerator.GeneratePaymentTransactionId(),
            Status = BookingStatus.Confirmed,
            ConfirmationEmailSent = false,
            BookingSource = BookingConstants.BookingSource,
            CreatedAt = now,
            UpdatedAt = now,
            Guests = request.Guests?.Select(guest => new BookingGuest
            {
                Sequence = guest.Sequence,
                FirstName = guest.FirstName.Trim(),
                LastName = guest.LastName.Trim(),
                Gender = guest.Gender,
                AgeGroup = guest.AgeGroup,
                CreatedAt = now,
                UpdatedAt = now,
            }).ToList() ?? [],
        };

        try
        {
            var created = await bookingRepository.CreateBookingAsync(
                booking,
                bookingSessionId,
                cancellationToken);

            var confirmationEmailSent = false;
            try
            {
                var bookingForEmail = await bookingRepository.FindDetailsByReferenceNumberAsync(
                    created.ReferenceNumber,
                    cancellationToken);

                if (bookingForEmail is not null)
                {
                    await confirmationNotifier.NotifyAsync(bookingForEmail, cancellationToken);
                    await bookingRepository.MarkConfirmationEmailSentAsync(
                        created.Id,
                        cancellationToken);
                    confirmationEmailSent = true;
                }
            }
            catch (Exception ex)
            {
                logger.LogWarning(
                    ex,
                    "Confirmation notification failed for {ReferenceNumber}",
                    created.ReferenceNumber);
            }

            return ServiceResult<CreateBookingResponseDto>.Ok(
                new CreateBookingResponseDto(
                    created.ReferenceNumber,
                    room.Name,
                    created.CheckInDate,
                    created.CheckOutDate,
                    created.NumberOfNights,
                    created.TotalPrice,
                    created.PaymentStatus,
                    created.Status,
                    confirmationEmailSent),
                201);
        }
        catch (ConflictException conflict)
        {
            return ServiceResult<CreateBookingResponseDto>.Fail(conflict.Message, conflict.StatusCode);
        }
    }

    public async Task<ServiceResult<MemberBookingsResponseDto>> ListByCurrentMemberAsync(
        CancellationToken cancellationToken = default)
    {
        if (!currentUser.IsAuthenticated || !currentUser.MemberId.HasValue)
        {
            return ServiceResult<MemberBookingsResponseDto>.Fail(
                ApplicationMessages.Authorization.AuthenticationRequired,
                401);
        }

        var bookings = (await bookingRepository.FindByMemberIdAsync(
                currentUser.MemberId.Value,
                cancellationToken))
            .Where(booking => BookingAuthorization.ShouldAppearInMemberHistory(booking, currentUser))
            .ToList();

        var summaries = bookings.Select(booking => new MemberBookingSummaryDto(
            booking.ReferenceNumber,
            booking.Room.Name,
            booking.CheckInDate,
            booking.CheckOutDate,
            booking.AdultCount,
            booking.ChildCount,
            booking.InfantCount,
            booking.Status,
            booking.TotalPrice)).ToList();

        return ServiceResult<MemberBookingsResponseDto>.Ok(
            new MemberBookingsResponseDto(summaries));
    }

    public async Task<ServiceResult<object>> GetByReferenceNumberAsync(
        string referenceNumber,
        CancellationToken cancellationToken = default)
    {
        var booking = await bookingRepository.FindDetailsByReferenceNumberAsync(
            referenceNumber,
            cancellationToken);

        if (booking is null)
        {
            return ServiceResult<object>.Fail(ApplicationMessages.Booking.NotFound, 404);
        }

        var viewDenied = BookingAuthorization.GetViewDeniedMessage(booking, currentUser);
        if (viewDenied is not null)
        {
            var statusCode =
                viewDenied == ApplicationMessages.Authorization.AuthenticationRequired
                    ? 401
                    : 403;

            return ServiceResult<object>.Fail(viewDenied, statusCode);
        }

        return ServiceResult<object>.Ok(MapBookingDetails(booking));
    }

    public async Task<ServiceResult<object>> GetByReferenceForManageAsync(
        string referenceNumber,
        CancellationToken cancellationToken = default)
    {
        var booking = await bookingRepository.FindDetailsByReferenceNumberAsync(
            referenceNumber,
            cancellationToken);

        if (booking is null)
        {
            return ServiceResult<object>.Fail(ApplicationMessages.Booking.NotFound, 404);
        }

        var manageDenied = BookingAuthorization.GetManageViewDeniedMessage(booking, currentUser);
        if (manageDenied is not null)
        {
            var statusCode =
                manageDenied == ApplicationMessages.Authorization.AuthenticationRequired
                    ? 401
                    : 403;

            return ServiceResult<object>.Fail(manageDenied, statusCode);
        }

        return ServiceResult<object>.Ok(MapBookingDetails(booking));
    }

    private static object MapBookingDetails(BookingEntity booking) =>
        new
        {
            booking.ReferenceNumber,
            booking.RoomId,
            roomName = booking.Room.Name,
            booking.FirstName,
            booking.LastName,
            booking.ContactEmail,
            booking.AddressLine1,
            booking.AddressLine2,
            booking.City,
            booking.State,
            booking.ZipCode,
            booking.Country,
            booking.AdultCount,
            booking.ChildCount,
            booking.InfantCount,
            booking.PetCount,
            booking.CheckInDate,
            booking.CheckOutDate,
            booking.NumberOfNights,
            booking.PricePerNight,
            booking.DiscountAmount,
            booking.TaxAmount,
            booking.TotalPrice,
            booking.Status,
            booking.PaymentStatus,
            booking.ConfirmationEmailSent,
            booking.SpecialRequests,
            guests = booking.Guests.OrderBy(g => g.Sequence).Select(g => new
            {
                g.Sequence,
                g.FirstName,
                g.LastName,
                g.AgeGroup,
            }),
        };

    public async Task<ServiceResult<CancelBookingResponseDto>> CancelAsync(
        string referenceNumber,
        CancelBookingRequestDto? request,
        CancellationToken cancellationToken = default)
    {
        var existing = await bookingRepository.FindByReferenceNumberAsync(referenceNumber, cancellationToken);

        if (existing is null)
        {
            return ServiceResult<CancelBookingResponseDto>.Fail(
                ApplicationMessages.Booking.NotFound,
                404);
        }

        var accessDenied = BookingAuthorization.GetAccessDeniedMessage(existing, currentUser);
        if (accessDenied is not null)
        {
            return ServiceResult<CancelBookingResponseDto>.Fail(accessDenied, 403);
        }

        if (existing.Status == BookingStatus.Cancelled)
        {
            return ServiceResult<CancelBookingResponseDto>.Fail(
                ApplicationMessages.Booking.AlreadyCancelled,
                400);
        }

        if (existing.CheckInDate.Date <= DateTime.UtcNow.Date)
        {
            return ServiceResult<CancelBookingResponseDto>.Fail(
                ApplicationMessages.Booking.CannotCancelOnOrAfterCheckIn,
                400);
        }

        var cancelled = await bookingRepository.CancelByReferenceNumberAsync(
            referenceNumber,
            request?.CancellationReason,
            cancellationToken);

        return ServiceResult<CancelBookingResponseDto>.Ok(
            new CancelBookingResponseDto(
                cancelled.ReferenceNumber,
                cancelled.Status,
                cancelled.PaymentStatus,
                ApplicationMessages.Booking.Cancelled,
                cancelled.UpdatedAt));
    }

    public async Task<ServiceResult<ModifyBookingResponseDto>> ModifyAsync(
        string referenceNumber,
        ModifyBookingRequestDto request,
        CancellationToken cancellationToken = default)
    {
        var booking = await bookingRepository.FindForUpdateByReferenceNumberAsync(
            referenceNumber,
            cancellationToken);

        if (booking is null)
        {
            return ServiceResult<ModifyBookingResponseDto>.Fail(
                ApplicationMessages.Booking.NotFound,
                404);
        }

        var accessDenied = BookingAuthorization.GetAccessDeniedMessage(booking, currentUser);
        if (accessDenied is not null)
        {
            return ServiceResult<ModifyBookingResponseDto>.Fail(accessDenied, 403);
        }

        if (booking.Status != BookingStatus.Confirmed)
        {
            return ServiceResult<ModifyBookingResponseDto>.Fail(
                ApplicationMessages.Booking.OnlyConfirmedCanBeModified,
                400);
        }

        if (booking.CheckInDate.Date < DateTime.UtcNow.Date)
        {
            return ServiceResult<ModifyBookingResponseDto>.Fail(
                ApplicationMessages.Booking.CannotModifyAfterCheckInPassed,
                400);
        }

        DateTime checkIn;
        DateTime checkOut;

        if (request.CheckInDate is not null)
        {
            var parsed = DateHelpers.ParseDateOnly(request.CheckInDate);
            if (parsed is null)
            {
                return ServiceResult<ModifyBookingResponseDto>.Fail(
                    ApplicationMessages.Booking.InvalidCheckInDate,
                    400);
            }

            checkIn = parsed.Value;
        }
        else
        {
            checkIn = booking.CheckInDate;
        }

        if (request.CheckOutDate is not null)
        {
            var parsed = DateHelpers.ParseDateOnly(request.CheckOutDate);
            if (parsed is null)
            {
                return ServiceResult<ModifyBookingResponseDto>.Fail(
                    ApplicationMessages.Booking.InvalidCheckOutDate,
                    400);
            }

            checkOut = parsed.Value;
        }
        else
        {
            checkOut = booking.CheckOutDate;
        }

        var stayDateValidation = DateHelpers.ValidateStayDateRange(
            checkIn.ToString("yyyy-MM-dd"),
            checkOut.ToString("yyyy-MM-dd"));

        if (!stayDateValidation.IsValid)
        {
            return ServiceResult<ModifyBookingResponseDto>.Fail(
                stayDateValidation.ErrorMessage!,
                400);
        }

        var adultCount = request.AdultCount ?? booking.AdultCount;
        var childCount = request.ChildCount ?? booking.ChildCount;
        var infantCount = request.InfantCount ?? booking.InfantCount;
        var petCount = request.PetCount ?? booking.PetCount;
        var totalGuests = adultCount + childCount + infantCount;
        var room = booking.Room;

        if (totalGuests > room.MaxGuests)
        {
            return ServiceResult<ModifyBookingResponseDto>.Fail(
                ApplicationMessages.Booking.GuestCountExceedsCapacity,
                400);
        }

        if (!room.PetsAllowed && petCount > 0)
        {
            return ServiceResult<ModifyBookingResponseDto>.Fail(
                ApplicationMessages.Booking.PetsNotAllowed,
                400);
        }

        if (room.PetsAllowed && petCount > 2)
        {
            return ServiceResult<ModifyBookingResponseDto>.Fail(
                ApplicationMessages.Booking.MaxPetsExceeded,
                400);
        }

        booking.CheckInDate = checkIn;
        booking.CheckOutDate = checkOut;
        booking.AdultCount = adultCount;
        booking.ChildCount = childCount;
        booking.InfantCount = infantCount;
        booking.PetCount = petCount;

        if (request.SpecialRequests is not null)
        {
            booking.SpecialRequests = request.SpecialRequests.Trim();
        }

        if (booking.BookingType != BookingType.Member &&
            !string.IsNullOrWhiteSpace(request.ContactEmail))
        {
            booking.ContactEmail = request.ContactEmail.Trim();
        }

        booking.NumberOfNights = DateHelpers.CalculateNumberOfNights(
            booking.CheckInDate,
            booking.CheckOutDate);
        booking.TaxAmount = decimal.Round(
            booking.PricePerNight * booking.NumberOfNights * BookingConstants.TaxRate,
            2);
        booking.TotalPrice = decimal.Round(
            booking.PricePerNight * booking.NumberOfNights - booking.DiscountAmount + booking.TaxAmount,
            2);

        try
        {
            var updated = await bookingRepository.ModifyBookingAsync(booking, cancellationToken);

            return ServiceResult<ModifyBookingResponseDto>.Ok(
                new ModifyBookingResponseDto(
                    updated.ReferenceNumber,
                    room.Name,
                    updated.CheckInDate,
                    updated.CheckOutDate,
                    updated.NumberOfNights,
                    updated.AdultCount,
                    updated.ChildCount,
                    updated.InfantCount,
                    updated.PetCount,
                    updated.TotalPrice,
                    updated.Status,
                    updated.PaymentStatus,
                    ApplicationMessages.Booking.Updated));
        }
        catch (ConflictException conflict)
        {
            return ServiceResult<ModifyBookingResponseDto>.Fail(conflict.Message, conflict.StatusCode);
        }
    }

    private string? ValidateMemberContactMatchesAccount(CreateBookingRequestDto request)
    {
        if (string.IsNullOrWhiteSpace(currentUser.Email) ||
            !string.Equals(
                request.ContactEmail.Trim(),
                currentUser.Email,
                StringComparison.OrdinalIgnoreCase))
        {
            return ApplicationMessages.Booking.MemberContactMustMatchAccount;
        }

        if (string.IsNullOrWhiteSpace(currentUser.FirstName) ||
            string.IsNullOrWhiteSpace(currentUser.LastName))
        {
            return ApplicationMessages.Booking.MemberContactMustMatchAccount;
        }

        if (!string.Equals(request.FirstName.Trim(), currentUser.FirstName, StringComparison.OrdinalIgnoreCase) ||
            !string.Equals(request.LastName.Trim(), currentUser.LastName, StringComparison.OrdinalIgnoreCase))
        {
            return ApplicationMessages.Booking.MemberContactMustMatchAccount;
        }

        return null;
    }

}
